src/libsyntax/parse/parser.rs

```diff
@@ -3755,6 +3737,18 @@ impl<'a> Parser<'a> {
         self.look_ahead(1, |t| t.is_ident() && !t.is_any_keyword())
     }
 
+    fn is_defaultness(&self) -> bool {
+        // `pub` is included for better error messages
+        self.token.is_keyword(keywords::Default) &&
+        self.look_ahead(1, |t| t.is_keyword(keywords::Impl) ||
+                        t.is_keyword(keywords::Const) ||
+                        t.is_keyword(keywords::Fn) ||
+                        t.is_keyword(keywords::Unsafe) ||
+                        t.is_keyword(keywords::Extern) ||
+                        t.is_keyword(keywords::Type) ||
+                        t.is_keyword(keywords::Pub))
+    }
```

Comment:
Could you do something like this instead, to keep `default` in the expected token set
```rust
fn eat_defaultness(&self) -> bool {
    let is_defaultness = self.is_defaultness();
    if is_defaultness {
        self.bump();
    } else {
        self.expected_tokens.push(/*default keyword token*/);
    }
    is_defaultness
}
```

All this logic can probably be just inlined into `parse_defaultness` as well.